### DISCUSSION TITLE
How do I use Sveltekit promise streaming with PocketBase?

### DISCUSSION THREAD

**Discussion by tomikaka22** [pocketbase/pocketbase#2870] [Q&A] [answered]
I would like to stream big PocketBase queries in my SvelteKit app, to improve perceived performance.

However SvelteKit needs everything in it's load functions to be serializable. By default, PocketBase returns non serializable promises.

Currently I wrap all my PocketBase responses in structuredClone(). I have to await them to do so. But if I understand correctly, awaiting them server-side negates all the benefits of streaming. 

How can I stream long PocketBase queries with SvelteKit?

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure if that's what you are trying to do, but you should be able to return a `Promise` with converted `structuredClone` data without `await` by chaining `.then((data) => structuredClone(data /* or data.items, etc. */))`.

If you are asking whether it is possible to pass the `load` method custom `fetch` implementation to the JS SDK request, then No, at the moment this is not possible (_see also https://github.com/pocketbase/js-sdk/issues/207; eventually I may also reconsider the custom `Record` class wrapping and return directly POJOs so that `structuredClone` may not be necessary in the future_).

**Reply by tomikaka22** [reply]
I tried 

```
const records = locals.pb.collection('termekek').getFullList(1, {
	sort: '-created'
});

return {
	streamed: {
		'records': records.then((data) => structuredClone(data))
	}
}
```

But it throws an error: `DOMException [DataCloneError]: #<Promise> could not be cloned.`

**Reply by ganigeorgiev** [reply]
It works fine for me when following the example from https://kit.svelte.dev/docs/load#streaming-with-promises and pasting it in a [SvelteLab REPL](https://www.sveltelab.dev/#code=N4IgdADgTgpgLnAljKUDGIBcoBmiA2MWoaA9gHZwyUDOxIA1ouQCZYg0BuM%2BV%2BAhgCMAtAFdEADn4gANCEGicOFFgDaARgBMAZhnqADDICcM7QBY96gOyWArHv3qZEi0ZtGJd0xduftmywA2Swt1awcnM1DDE3MvA3swwwNdA08DJwknMJsw7PVgjO8ZX1MAsML1UPCiqIdjYoqQ41z9RP10tstk-V1A3LCHEwMTN2LS-0aDS3yepzi66djq8qqHROmmiVb25IKHCziJgLMTM09zmUWYqfKOy1bUrqSvI78Ao3mV5qKmv%2BfHM0zBs7ukar11t0rhsNp1svd-l9jLprg04mFQQ8HE8NoYJE8nGMwq5hvoLIFOhtqvY3mVjEizME4kYUdE0dV8nCfmsEg45q8fO96XodHT0fo7u0uWEaYKxRYzCdKrCIg4wYVAQtDgF0YMMUFIXkoTLxkKtLpJrqPgNqariab5fF1NKEfsinFJloaeVNPYWfUvQ4ALoAXzDcgg-DQDH4AHMYGAAFY0Cj0PCEehkSjUOB0bCMZhsTAcbi8HhCMSSaRyBRKFSYDSipZTGKtUlI45TFW-V0a%2BGeMYeaHGAcDdp%2B5ZXNmTsLSnnqcczKG0y0Ki7BS6Mq6T1EzrQ-Z7OloG91yz1u%2BetSLTqaFTT5O4doXi4Z6-R7VpcrX1PcTg8gk8FwdT0mxuSchxyQ0EXfYCdWqSkT01D4wR2NU%2BWKXd2XQ34jxXODLAQxE6QMflZwwm891WaJOSg-JJ07XVCLdJDLCou0j15HCdwoxoJSXe0RmPJo8NuA98kqDV2nybVh2BVi0MglJDWmOSTSKIk1iMUkFV2ZSZK3SY5NU9RX2iDY%2BzpAyFR45ltNVLSHBsETdW7QE0kQyJ2lQ4TVkSe8oQcy0-DNICyJ7dVVTkhyDGo4xSSck57GM5U4v4zY%2BM4iFDI3VoUuivV0slbErg1FKjRecUGkDTDxVIo8KXYx9jRgnCYlJKSBRKM1QIaSdAltMK3N7O19BSsx3FaVt5KI%2B0ooGAZKifR0jB8Qwt03Dc-Gsn8sP6p0XQinsxtabz9mc1wFTOYJbDW9cSnwzDKKXaL9CvA0hw0jrzuMNcSu3EoUWuwxv2bJ6BLci9JPYw4z3wlaSjW66N028idrBnkIYk2igQA9zFIhTiVJhAiRsOzVYeKeGbr%2B04rju2mQbArsSfCjzYMpnwlTpmmtqnNHmcG-s2aiuz1Pi9nmR8SJrq52wGRs%2BDlIs76qdl6WShhnbA02X1gzDEM5BodB6BYRBYDQOBSCgABPeh%2BAgCAwAACzgABbfA0wIIh8yzKhaHoJhWHYLgeD4CtxCkWR5EUZQoDUQJgb6TwrBMfpnAsCQTAkQxAliO5qg1UlPECO4ZATg5EKL4j7g0no%2BiRVrihLnby5i%2BzJrL0vRjLnp23ehLUrbjSBsxsuG5qZjEpKJl9JsZvQZ7oZ2OCMZzymwSikCcfCIh19Vmkg0PQ%2BV9ZjZl4mi3qZ2uw5jyXhOqEJezzLGvkjEM1wy567tExxZoayfhKdZGFd1Id20K0Bo5dxogMBESGC88eiQJiGCe%2BIp5h3HXvsMYxlXJCxmq9JWw59TgzwYMOBa1dIaRXnPBuZI0pzk1gg3qi836ExCuZNBClASX1qspXGNdcj%2BU7PlU%2BM1NgXllJ3Uwv9BYKWYoAwhW4IJvmSP5cBi9oGsMPMXb%2BUD3AoOwv5JheiQHMSrgg0M4YQBQFIKIKgeZQCm3NpbG29AADUAhra2LgGAEOZZPYZh9hQP2uYA6FmDqWMOIgI7VmjnWOODZW6hVdOJSyX89ZWLcZGeMviUDcCgEmBxIB0zexIMEnMRTA5FhLKHcs0SqxR1rLHNQmU95pQLnSbOz1xY4WCIEah8QWIGDYuVfQedRJGhem9JoH0ZHkwDKRTQex97ciZHcVYAw2qIQ2KfT0KShIwTMGtdBIoni9RPsaUKZ1VjITHsRdKdFOl7EJD00e-ShLJMiMDaIHS-j7hNKUaBmj-lkNeak2Zgz1aEwZPCNGFzz5XLWWJQeBVoh7G9Hc88tylHwuglyfKUNyoqW%2BdNAqHCTQLChQTUeVgPxcOToSxFVxjkLLFDqdlyK-k2mXD6M5loOX8rZfc0RgJOxdIEm%2BV%2Bny-okKdKQrkRz-pmGvPUQVaqBUcrUheWKcC%2B5v07OaVVGqhXqpNZq2%2BnLRUhR1fCjG6KZUWsboqxVb8qHHmVXSS%2BzdJhWAzroFOMgrB3X1MPSFzLA1%2BsDSYINDryUY1xMVTKPovnoTddgyIGajVmuzSBdoOafT2FsKSEUthLEGxAFkuMCY-FUACaUkAvsKlhKDsWGtdTKyRxrDHesqgkl-nBrjfYTDTWfyuCcDle0-j%2BWIUM0C7Q2lEo6Yq7QV0rhOSxloB87EaRnHVl0uBBIy6jn7iWyyJwIEjqaloZZ7TD6Yl6dXTBAyi31BMSafKZlSV9NLtMFwaUBnuSxWlONhbOgsP3MO41IELR%2BgGKMIBj6PkDK3FqyoOq2wjURu-S12RdbTEvfmwNi0Dk9ATUpdawD8NQd5V1bDjdzzHrxuCXDeaCMEcNWROBKFIoajnHXOjeHgbUdObRhyaa7J7IPhKqTgnCPsdAlJ%2BcB8UNjO3V4A9dpZPyZRBNF%2BBzyWybfRBzQZb9blrbWALMeBYyFLrZmcp-t8xVIibUgQ9TO1xOaQ2VppLFNIvIR81YFl7R7JGZKsUqcBocX-mfUaXhoEBY0kF6Gw4NILSXCukoxaVI2EVGOz1vzp3-NnTiJexDUUGmXZF-apNYvHUcoQzLHqtxpyJT0IqSlAyGXPT%2BW19qfIopFR1rKAQeE0ebAYBDBqerdyBQPRLAxks4R0h89LRpzwivSGo3yWJysrIlSlc8OyytDIhnsJiHJmp6vEWNGSakhonLbn8yoO3AW5ZOoosq0XSEWRUw-NdN8Xnxd066xbZ81hzV2xeR7GUKuTFXvew8HCVPMLm3SKwTxNVFfyNyyTlyLtpV7uxPYSLCYYoRako7UObm3vwUJ29IF8g4-uR1Z7XDUEbZ5ceKnM6gdJshIFK94iCdtZO6I-YdQecmUIS9O7BV0vng1LFeHHckvg4VBqVH73OmHvdPnQCL2vBvYB5Nz7NXWZET%2B%2BxC4HdJmviSiDmCauxG2lCwzu4trlcdcKrT1TbroVc8HAYxS1ySOBfV-FzX3cFtfpS8I1XYOXfho9RN2TgYX3pU0zexdd7sJDWtLt4a8O%2BsOGG7oDP%2Bgy1yE4IgKglmKDWds-mEp9nsyOdAM51tkT20xMad2hJGgAQLoqwNw1UuDqxYQ7m6uYX4TjMy5Oq5E-8H1dN4THKf9yWahMBXqCw%2B-PEVFHs0iGVDz8lawVdrUELQFpn9yOfEXN9s0y7vh5Ips8j9D1o%2BEBf8YRWPyNN7ufgMENtfuGhxjcDOtxp1mfkbkKJ8ObjFvgghs6oSGctVE4BXqZkAA) (_it uses `JSON.parse/stringify` because `structuredClone` doesn't seem available in the repl_).

**Reply by tomikaka22** [reply]
Oh my bad! I forgot to look at the +page.svelte file where I attempted to clone my returned data for a completely different reason. That's why it was complaining. It appears to be working. 

Thank you for your patience and quick replies!

**Comment by lucasyoboy** [comment]
Hello. How then could I catch the error?

**Reply by ganigeorgiev** [reply]
@lucasyoboy This is an old discussion and many of the above comments are no longer valid (_with the latest SDK version you can also pass the SvelteKit fetch implementation - [Custom request options](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#custom-request-options)_).

For general error handling, see https://github.com/pocketbase/js-sdk?tab=readme-ov-file#error-handling.

If you need further help please open a new Q&A discussion with more details about your use case and what you are trying to do.

**Reply by lucasyoboy** [reply]
you are quick. Thanks.

**Reply by lucasyoboy** [reply]
but how would the svelteKit fetch implementation work?
